src/mocha/utils.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+'use strict';
+
+/* eslint-env browser, worker, mocha */
+
+const mochaUtils = {
+  startMochaTest: () => {
+    let passedTests = [];
+    let failedTests = [];
+
+    mocha.checkLeaks();
+
+    var runResults = window.mocha.run();
+    // pass, fail and end events allow up to capture results and
+    // determine when to publish test results
+    runResults.on('pass', test => {
+      passedTests.push(mochaUtils.getFriendlyTestResult(test));
+    })
+    .on('fail', test => {
+      failedTests.push(mochaUtils.getFriendlyTestResult(test));
+    })
+    .on('end', () => {
+      mochaUtils.publishTestResults(passedTests, failedTests);
+    });
+
+    // No tests so end won't be called
+    if (mocha.suite.suites.length === 0) {
+      console.warn('No tests found.');
+      mochaUtils.publishTestResults(passedTests, failedTests);
+    }
+  },
+
+  getFriendlyTestResult: testResult => {
+    const friendlyResult = {
+      parentTitle: testResult.parent.title,
+      title: testResult.title,
+      state: testResult.state
+    };
+
+    if (testResult.err) {
+      friendlyResult.errMessage = testResult.err.message;
+    }
+
+    return friendlyResult;
+  },
+
+  publishTestResults: (passedTests, failedTests) => {
+    window.testsuite = {
+      testResults: {
+        passed: passedTests,
+        failed: failedTests
+      }
+    };
+  },
+
+  prettyPrintErrors: (identifier, testResults) => {
+    // Print test failues
+    if (testResults.failed.length > 0) {
+      const failedTests = testResults.failed;
+      let errorMessage = 'Issues in ' + identifier + '.\n\n' + identifier +
+        ' had ' + testResults.failed.length + ' test failures.\n';
+      errorMessage += '------------------------------------------------\n';
+      errorMessage += failedTests.map((failedTest, i) => {
+        return `[Failed Test ${i + 1}]\n` +
+               `    - ${failedTest.parentTitle} > ${failedTest.title}\n` +
+               `        ${failedTest.errMessage}\n`;
+      }).join('\n');
+      errorMessage += '------------------------------------------------\n';
+      return errorMessage;
+    }
+    return null;
+  }
+};
+
+/* eslint-disable no-negated-condition */
+if (typeof window !== 'undefined') {
```

Comment:
Damn it. Thought I got them all.

On Tue, 5 Apr 2016, 19:32 Jeffrey Posnick, <notifications@github.com> wrote:

> In src/mocha/utils.js
> <https://github.com/GoogleChrome/sw-testing-helpers/pull/6#discussion_r58591666>
> :
>
> > +        ' had ' + testResults.failed.length + ' test failures.\n';
> > +      errorMessage += '------------------------------------------------\n';
> > +      errorMessage += failedTests.map((failedTest, i) => {
> > +        return `[Failed Test ${i + 1}]\n` +
> > +               `    - ${failedTest.parentTitle} > ${failedTest.title}\n` +
> > +               `        ${failedTest.errMessage}\n`;
> > +      }).join('\n');
> > +      errorMessage += '------------------------------------------------\n';
> > +      return errorMessage;
> > +    }
> > +    return null;
> > +  }
> > +};
> > +
> > +/* eslint-disable no-negated-condition */
> > +if (typeof window !== 'undefined') {
>
> See my other comment about assuming that window.mocha will exist, which
> implies that window is always going to be defined.
>
> —
> You are receiving this because you authored the thread.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleChrome/sw-testing-helpers/pull/6/files/ae49a2c1893389c457b1bce0f9cbd620f05dc8d8#r58591666>
>